hudi-client/src/main/java/org/apache/hudi/metadata/HoodieMetadataImpl.java

```diff
@@ -0,0 +1,1064 @@
+      } else {
+        Arrays.stream(t._2).forEach(status -> {
+          String filename = status.getPath().getName();
+          if (filename.equals(HoodiePartitionMetadata.HOODIE_PARTITION_METAFILE)) {
+            return;
+          }
+          HoodieWriteStat writeStat = new HoodieWriteStat();
+          writeStat.setPath(partition + Path.SEPARATOR + filename);
+          writeStat.setPartitionPath(partition);
+          writeStat.setTotalWriteBytes(status.getLen());
+          metadata.addWriteStat(partition, writeStat);
+        });
+      }
+      stats[0] += t._2.length;
+    });
+
+    LOG.info("Committing " + partitionFileList.size() + " partitions and " + stats[0] + " files to metadata");
+    update(metadata, createInstantTime);
+  }
+
+  /**
+   * Sync the Metadata Table from the instants created on the dataset.
+   *
+   * @param datasetMetaClient {@code HoodieTableMetaClient} for the dataset
+   */
+  private void syncFromInstants(HoodieTableMetaClient datasetMetaClient) throws IOException {
+    ValidationUtils.checkState(!readOnly, "Metadata table cannot be synced in readonly mode");
+
+    List<HoodieInstant> instantsToSync = findInstantsToSync(datasetMetaClient);
+    if (instantsToSync.isEmpty()) {
+      return;
+    }
+
+    LOG.info("Syncing " + instantsToSync.size() + " instants to metadata table: " + instantsToSync);
+
+    // Read each instant in order and sync it to metadata table
+    final HoodieActiveTimeline timeline = datasetMetaClient.getActiveTimeline();
+    for (HoodieInstant instant : instantsToSync) {
+      LOG.info("Syncing instant " + instant + " to metadata table");
+
+      switch (instant.getAction()) {
+        case HoodieTimeline.CLEAN_ACTION: {
+          // CLEAN is synced from the
+          // - inflight instant which contains the HoodieCleanerPlan, or
+          // - complete instant which contains the HoodieCleanMetadata
+          try {
+            HoodieInstant inflightCleanInstant = new HoodieInstant(true, instant.getAction(), instant.getTimestamp());
+            ValidationUtils.checkArgument(inflightCleanInstant.isInflight());
+            HoodieCleanerPlan cleanerPlan = CleanerUtils.getCleanerPlan(datasetMetaClient, inflightCleanInstant);
+            update(cleanerPlan, instant.getTimestamp());
+          } catch (HoodieIOException e) {
+            HoodieInstant cleanInstant = new HoodieInstant(false, instant.getAction(), instant.getTimestamp());
+            ValidationUtils.checkArgument(cleanInstant.isCompleted());
+            HoodieCleanMetadata cleanMetadata = CleanerUtils.getCleanerMetadata(datasetMetaClient, cleanInstant);
+            update(cleanMetadata, instant.getTimestamp());
+          }
+          break;
+        }
+        case HoodieTimeline.DELTA_COMMIT_ACTION:
+        case HoodieTimeline.COMMIT_ACTION:
+        case HoodieTimeline.COMPACTION_ACTION: {
+          ValidationUtils.checkArgument(instant.isCompleted());
```

Comment:
we are not filtering for completed compactions right?  in fact, you wont find an instant with `compaction` as action type in compelted, since we make that a commit action eventually.

Reply:
Yes, HoodieTimeline.COMPACTION_ACTION is simply for completion purpose. Do you propose I remove this or add an exception or code comment?